File: factory_method_pattern/fmp.cpp
// C++ Implementation
#include <iostream>
#include <string>
#include <memory>

class Animal {
public:
    virtual std::string speak() = 0;
    virtual ~Animal() = default;
};

class Dog : public Animal {
public:
    std::string speak() override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    std::string speak() override {
        return "Meow!";
    }
};

class AnimalFactory {
public:
    std::unique_ptr<Animal> create_animal(const std::string& animal_type) {
        if (animal_type == "dog") {
            return std::make_unique<Dog>();
        } else if (animal_type == "cat") {
            return std::make_unique<Cat>();
        }
        return nullptr;
    }
};

// Example usage
int main() {
    AnimalFactory factory;
    auto dog = factory.create_animal("dog");
    auto cat = factory.create_animal("cat");
    
    std::cout << "Dog says: " << dog->speak() << std::endl;
    std::cout << "Cat says: " << cat->speak() << std::endl;
    
    return 0;
}

File: factory_method_pattern/fmp.kt
abstract class Animal {
  abstract fun speak(): String
}

class Dog : Animal() {
  override fun speak(): String {
      return "Woof!"
  }
}

class Cat : Animal() {
  override fun speak(): String {
      return "Meow!"
  }
}

class AnimalFactory {
  fun createAnimal(animalType: String): Animal? {
      return when (animalType) {
          "dog" -> Dog()
          "cat" -> Cat()
          else -> null
      }
  }
}

// Example usage
fun main() {
  val factory = AnimalFactory()
  val dog = factory.createAnimal("dog")
  val cat = factory.createAnimal("cat")
  
  println("Dog says: ${dog?.speak()}")
  println("Cat says: ${cat?.speak()}")
}